reports-scheduler/src/main/java/com/amazon/opendistroforelasticsearch/reportsscheduler/rest/handler/ReportsJobActionHandler.java

```diff
@@ -0,0 +1,134 @@
+package com.amazon.opendistroforelasticsearch.reportsscheduler.rest.handler;
+
+import com.amazon.opendistroforelasticsearch.jobscheduler.spi.JobExecutionContext;
+import com.amazon.opendistroforelasticsearch.jobscheduler.spi.utils.LockService;
+import com.amazon.opendistroforelasticsearch.reportsscheduler.job.ReportsSchedulerJobRunner;
+import com.amazon.opendistroforelasticsearch.reportsscheduler.job.parameter.JobParameter;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.elasticsearch.action.ActionListener;
+import org.elasticsearch.action.DocWriteResponse;
+import org.elasticsearch.action.delete.DeleteRequest;
+import org.elasticsearch.action.delete.DeleteResponse;
+import org.elasticsearch.action.search.SearchRequest;
+import org.elasticsearch.action.search.SearchResponse;
+import org.elasticsearch.client.node.NodeClient;
+import org.elasticsearch.cluster.service.ClusterService;
+import org.elasticsearch.common.xcontent.json.JsonXContent;
+import org.elasticsearch.index.query.QueryBuilders;
+import org.elasticsearch.rest.BytesRestResponse;
+import org.elasticsearch.rest.RestChannel;
+import org.elasticsearch.rest.RestResponse;
+import org.elasticsearch.rest.RestStatus;
+import org.elasticsearch.search.SearchHit;
+import org.elasticsearch.search.SearchHits;
+import org.elasticsearch.search.builder.SearchSourceBuilder;
+import org.elasticsearch.search.sort.FieldSortBuilder;
+import org.elasticsearch.search.sort.SortOrder;
+
+import java.util.Locale;
+
+public class ReportsJobActionHandler extends AbstractActionHandler {
+    private final LockService lockService;
+    private static final Logger log = LogManager.getLogger(ReportsJobActionHandler.class);
+    private final Long LOCK_DURATION_SECONDS = 300L;
+
+    /**
+     * Constructor function.
+     *
+     * @param client         ES node client that executes actions on the local node
+     * @param channel        ES channel used to construct bytes / builder based outputs, and send responses
+     * @param clusterService to be added
+     */
+    public ReportsJobActionHandler(NodeClient client, RestChannel channel, ClusterService clusterService) {
+        super(client, channel);
+        this.lockService = new LockService(client, clusterService);
+    }
+
+
+    public void getJob() {
+        // get all jobs from queue_index
+        // return 404 if no jobs in the queue, or the queue doesn't exist at all
+        SearchRequest searchRequest = new SearchRequest(ReportsSchedulerJobRunner.JOB_QUEUE_INDEX_NAME);
+        SearchSourceBuilder searchSourceBuilder = new SearchSourceBuilder();
+        searchSourceBuilder.query(QueryBuilders.matchAllQuery());
+
+        // sort the document
+        searchSourceBuilder.sort(new FieldSortBuilder("enqueue_time").order(SortOrder.DESC));
```

Comment:
sorting may lead to clashing. randomizing would reduce clashing and faster access of jobs (instead of serial waiting for locks)

Reply:
please review the latest changes from all commits